dune_client/client_async.py

```diff
@@ -0,0 +1,152 @@
+""""
+Async Dune Client Class responsible for refreshing Dune Queries
+Framework built on Dune's API Documentation
+https://duneanalytics.notion.site/API-Documentation-1b93d16e0fa941398e15047f643e003a
+"""
+import asyncio
+from typing import Any
+
+from aiohttp import ClientSession, ClientResponse, ContentTypeError, TCPConnector, ClientTimeout
+
+from dune_client.base_client import BaseDuneClient
+from dune_client.models import (
+    ExecutionResponse,
+    DuneError,
+    ExecutionStatusResponse,
+    ResultsResponse,
+    ExecutionState,
+)
+
+from dune_client.query import Query
+
+
+# pylint: disable=duplicate-code
+class AsyncDuneClient(BaseDuneClient):
+    """
+    An asynchronous interface for Dune API with a few convenience methods
+    combining the use of endpoints (e.g. refresh)
+    """
+    _session = None
+    _connection_limit = 3
+
+    def __init__(
+            self,
+            api_key: str,
+            connection_limit: int = 3
+    ):
+        """
+        api_key - Dune API key
+        connection_limit - number of parallel requests to execute.
+        For non-pro accounts Dune allows only up to 3 requests but that number can be increased.
+        """
+        super().__init__(api_key=api_key)
+        self._connection_limit = connection_limit
+        self._session = self._create_session()
+
+    def _create_session(self):
+        conn = TCPConnector(limit=self._connection_limit)
+        return ClientSession(
+            connector=conn,
+            base_url=self.BASE_URL,
+            timeout=ClientTimeout(total=self.DEFAULT_TIMEOUT),
+        )
+
+    async def close_session(self):
+        """Closes client session"""
+        await self._session.close()
+
+    async def __aenter__(self):
+        self._session = self._create_session()
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        await self.close_session()
```

Comment:
Could you elaborate a bit on the meaning of these?

Reply:
Sure! It's for `async with` syntax. see https://peps.python.org/pep-0492/#asynchronous-context-managers-and-async-with.

Reply:
Nice thanks! I guess this is analogous to `with open(file...):` and does session closing instead of file closing etc... Makes perfect sense.